app/models/voucher.rb

```diff
@@ -0,0 +1,65 @@
+class Voucher < ApplicationRecord
+  ROLE_HASH = { tutor: 0, editor: 1, teacher: 2, speaker: 3 }.freeze
+  SPEAKER_EXPIRATION_DAYS = 30
+  TUTOR_EXPIRATION_DAYS = 14
+  DEFAULT_EXPIRATION_DAYS = 3
+
+  enum role: ROLE_HASH
+
+  belongs_to :lecture, touch: true
+  before_create :generate_secure_hash
+  before_create :add_expiration_datetime
+  before_create :ensure_no_other_active_voucher
+  before_create :ensure_speaker_vouchers_only_for_seminars
+  validates :role, presence: true
+
+  scope :active, lambda {
+                   where("expires_at > ? AND invalidated_at IS NULL",
+                         Time.zone.now)
+                 }
+
+  self.implicit_order_column = "created_at"
+
+  def self.roles_for_lecture(lecture)
+    return ROLE_HASH.keys if lecture.seminar?
+
+    ROLE_HASH.keys - [:speaker]
+  end
+
+  private
+
+    def generate_secure_hash
+      self.secure_hash = SecureRandom.hex(16)
+    end
+
+    def add_expiration_datetime
+      self.expires_at = created_at + expiration_days.days
+    end
+
+    def ensure_no_other_active_voucher
+      return unless lecture
```

Comment:
I am not really sure in which order the validations corresponding to
`belongs_to :lecture`
and
`before_create :ensure_no_other_active_voucher`
are evaluated. If they are run in the order that I just used, then everything should be fine without the `return unless lecture` statement because a missing lecture would result in an error that the lecture is missing (with a proper error message). If they are run the other way around and the lecture is missing, then without the `return unless lecture` statement there would be an exception of the form "undefined method voucher for nil" raised by the next line in `ensure_no_other_voucher`. With the `return unless lecture` statement, nothing is returned and the exception is raised by the validation for `belongs_to :lecture` (with the proper error message). So, yes, in that sense this is intended. But I am not overly attached to this, if you think this is not useful then we can take that line out.

Reply:
I'm not too sure either based on some digging in the documentation, so I posted [this question on StackOverflow](https://stackoverflow.com/q/78959006/9655481). Let's see if we get an answer there.

For now, we can leave the `return unless lecture` in here as safeguard.

Reply:
@fosterfarrell9 We got a nice answer [here](https://stackoverflow.com/a/78959761/9655481) that argues with the actual Rails soruce code. Based on this, I will just leave the `return unless lecture` in for safety, even though it's almost certain that we would not run into any problems if omitted.